app/src/main/java/io/stormbird/wallet/service/KeyService.java

```diff
@@ -0,0 +1,1412 @@
+    {
+        //check unlock status
+        try
+        {
+            KeyStore keyStore = KeyStore.getInstance(ANDROID_KEY_STORE);
+            keyStore.load(null);
+            SecretKey secretKey = (SecretKey) keyStore.getKey(currentKeyAddress, null);
+            String encryptedHDKeyPath = getEncryptedFilePath(currentKeyAddress);//getFilePath(context, currentKey + "hd");
+            boolean fileExists = new File(encryptedHDKeyPath).exists();
+            if (!fileExists || secretKey == null)
+            {
+                signCallback.GotAuthorisation(false);
+                return;
+            }
+            byte[] iv = readBytesFromFile(getFilePath(context, currentKeyAddress + "iv"));
+            Cipher outCipher = Cipher.getInstance(CIPHER_ALGORITHM);
+            final GCMParameterSpec spec = new GCMParameterSpec(128, iv);
+            outCipher.init(Cipher.DECRYPT_MODE, secretKey, spec);
+            signCallback.GotAuthorisation(true);
+            return;
+        }
+        catch (UserNotAuthenticatedException e)
+        {
+            checkAuthentication(Operation.CHECK_AUTHENTICATION);
+            return;
+        }
+        catch (KeyPermanentlyInvalidatedException | UnrecoverableKeyException e)
+        {
+            //see if we can automatically recover the key
+            keyFailure("Key created at different security level. Please re-import key");
+            e.printStackTrace();
+        }
+        catch (Exception e)
+        {
+            //some other error, will exit the recursion with bad
+            e.printStackTrace();
+        }
+
+        signCallback.GotAuthorisation(false);
+    }
+
+    private synchronized String unpackMnemonic(String keyAddr, Operation operation)
+    {
+        try
+        {
+            KeyStore keyStore = KeyStore.getInstance(ANDROID_KEY_STORE);
+            keyStore.load(null);
+            if (!keyStore.containsAlias(keyAddr))
+            {
+                keyFailure("Key not found in keystore. Re-import key.");
+                return "";
+            }
+
+            //create a stream to the encrypted bytes
+            FileInputStream encryptedHDKeyBytes = getEncryptedFileStream(keyAddr);
+
+            SecretKey secretKey = (SecretKey) keyStore.getKey(keyAddr, null);
+            boolean ivExists = new File(getFilePath(context, keyAddr + "iv")).exists();
+            byte[] iv = null;
+
+            if (ivExists)
+                iv = readBytesFromFile(getFilePath(context, keyAddr + "iv"));
+            if (iv == null || iv.length == 0)
+            {
+                keyFailure("Cannot setup wallet seed.");
+                return "";
+            }
+            Cipher outCipher = Cipher.getInstance(CIPHER_ALGORITHM);
+            final GCMParameterSpec spec = new GCMParameterSpec(128, iv);
+            outCipher.init(Cipher.DECRYPT_MODE, secretKey, spec);
+            CipherInputStream cipherInputStream = new CipherInputStream(encryptedHDKeyBytes, outCipher);
```

Comment:
If tag verification fails during decryption you simply print a stack trace. Presumably we want to do something else since this would mean that the stored key is destroyed (and possibly maliciously tampered with) and we would want to start a restore flow/notify the user.